### DISCUSSION TITLE
Scanner points - size with zoom, or with res only

### DISCUSSION THREAD

**Discussion by mefistotelis** [swfans/syndwarsfx#207] [Ideas]
I got frustrated by the small size of dots on minimap in higher resolutions. Time to do something with that.

But I wonder - should the dots be scaled only on resolution change, or should the size of dots also change with zoom level?
What do you think?

I think it should be bound to the size of terrain visible on minimap - so zooming in should enlarge all the dots.

**Comment by geist22** [comment]
Dots... do you mean the crosses for other people on the minimap?

Is there a way to change minimap zoom? I thought it was tied to the general zoom level which is tied to the active weapon by default. In the DOS version, changing from no weapon to LR Rifle, the crosses look like they're always the same size to me, 1 pixel on each side in 640x480, regardless of zoom.
My minimap in the port looks more zoomed in with higher resolution (the higher the resolution, the less terrain it shows per area in the map).

This isn't just related to the minimap, but what I would like to have is UI scale so the interface elements don't "shrink" with higher resolution but are always the same relative size on your screen.

**Comment by mefistotelis** [comment]
> Dots... do you mean the crosses for other people on the minimap?

Yes, the dots/crosses (switched every turn). And circles around targets. And circle expanding into dotted circle around your agents.
And an arc around closest target.

Some initial scaling is now available in latest build. In higher res the circles are larger, in 320x200 some are smaller than original.
(with linear scaling everything was reduced to 1 pixel, so I introduced some limits, but still allowed some reduction).

**Comment by mefistotelis** [comment]
> My minimap in the port looks more zoomed in with higher resolution (the higher the resolution, the less terrain it shows per area in the map).

You're right - it is roughly the same for 320x300 and 640x480, but then starts dropping. I didn't noticed that..

**Reply by mefistotelis** [reply]
Try it now - I changed how the scanner zoom is calculated:
[swars: Changed the way of computing scanner zoom](https://github.com/swfans/swars/commit/f62eb52c423cacba37d615e4880b881fc481be84)

**Reply by geist22** [reply]
Latest build crashes my game when I enter a mission - only the menus work.

**Reply by mefistotelis** [reply]
Just tried starting the game, entering new campaign and starting the first mission. It works for me.
Tried with both original and altered resolutions, with mine and non-exisiting login settings.

Can you check at which commit this started?

**Reply by geist22** [reply]
https://github.com/swfans/swars/commit/1de4549a4f191f880518d124c02fb5dc179e9bee still works for me so https://github.com/swfans/swars/commit/f62eb52c423cacba37d615e4880b881fc481be84 is the first that crashes.
I've tried 640x480, 320x200 from cmd and new game from the menu, everything crashes.

-edit-
I've found the cause, it's because I'm using edited values for zoom in rules.ini (173 for both `ZoomMin` and `ZoomMax`). It doesn't crash with default values.

**Reply by mefistotelis** [reply]
> 173 for both `ZoomMin` and `ZoomMax`

Oh right. If both values are the same then this division breaks math:
```
zoom += 128 * (user_zoom_max - get_unscaled_zoom(overall_scale)) / (get_overall_scale_max() - get_overall_scale_min());
```

**Reply by geist22** [reply]
I've checked it out using the default zoom levels now. There are several bugs:

1. If you zoom in too much (closer than a Minigun), NPCs change from crosses to squares in 640x480. It still works in 320x200, and in higher resolutions they look like circles instead. Edit: This also happens with LR Rifle zoom in a 1280x960 window (using 640x480 resolution), so it's not zoom alone that causes this. I don't really know if NPCs were meant to be circles instead and crosses are just a constraint of the resolution, though they've always been crosses in the original game.

![map](https://github.com/user-attachments/assets/8c259dc5-b470-442a-827c-86799ee50593)

2. Agents are now a circle made of dots rather than a closed circle, see image above.
3.  The zoom doesn't match the original, I've taked a few comparison screenshots:

noweapon DOS vs. noweapon port:
![nowep_dos](https://github.com/user-attachments/assets/66943905-1693-47e6-8cbc-e526b33d27a5) ![nowep_port](https://github.com/user-attachments/assets/22e57596-f03e-4995-8b15-98cdd8c46d2c)

LR Rifle DOS vs. LR Rifle port:
![LR_dos](https://github.com/user-attachments/assets/d3872acf-50a8-4b96-b974-accb8ea17473) ![LR_port](https://github.com/user-attachments/assets/bc30b04f-3fcd-461e-8ed1-c652908caa1a)

So the noweapon zoom is pretty close to the original, a little bit too zoomed in, while it gets worse the further you zoom out, with LR rifle being quite a bit too zoomed in.

**Comment by mefistotelis** [comment]
> If you zoom in too much (closer than a Minigun), NPCs change from crosses to squares in 640x480.

Yes, that's how i implemeted the circles.
diagonal=1 -> one pixel
diagonal=2 -> one pixel
diagonal=3 -> 3x3 cross
diagonal=4 -> 3x3 rectangle
diagonal=5+ -> more or less a circle, filled

> Agents are now a circle made of dots rather than a closed circle, see image above.

There's a second drawing function, for dotted circles - because the circles around agents were originally kind of dotted.
(I mean the largest circles - these are animations of 4 frames, with circle of increasing size)

Currently the size of circles changes not only with resolution, but also with zoom.

> The zoom doesn't match the original

The original game has different scanner zoom levels for 320x200 and 640x480. The lowres one reveals a lot less, while highres one reveals more than half of the level - neither looks ok for my taste. So instead of reproducing the original, I went with what I think looks better.

If you want to look at the circles in detail, coords for drawing the original largest red circle are in `circle_dots_sz8`. Original circles around targets are in `circle_line_sz7`. All the other circle arrays were added by me. Circles below the size of 5 don't have arrays - are drawn directly within function `SCANNER_draw_circle_line`.

**Reply by geist22** [reply]
I'd prefer if you kept things closer to the original. Replacing crosses with bigger squares in 640x480 looks like a visual downgrade to me, so I'd go with diagonal=4 -> 3x3 cross too.

I've always welcomed a more zoomed out minimap, as it warns you earlier of enemy groups/vehicles closing in on your position and you don't have to move the camera around so much to scan the map for enemies at the start of a mission. Though I also don't use any abnormally high (low) zoom levels outside of the original game's highest (LR Rifle). This should optimally emulate the original game and have an .ini setting to adjust for personal preference.

Circles are only really dotted around targets, the ones around your agents look closed:
![agent](https://github.com/user-attachments/assets/f8cedcd3-8ff9-4acf-8b9e-297661253306)
If they were dotted, the 3 pixel wide lines would probably have the middle pixel removed. Player units having a unique look different from targets also makes it easier to identify them at quick glance.

New bug: It looks like the minimap doesn't update for the zoom I've set in `rules.ini` and just uses noweapon zoom instead.

I did some testing, in 1280x960. For me, the circles are a bit too big. Them having more volume compared to the crosses makes the "pulse" quite irritating, which didn't really bother me before.

**Reply by mefistotelis** [reply]
> Circles are only really dotted around targets, the ones around your agents look closed

Oh you're right. In original game the circle size goes 0, 1, 3, 5. I somehow thought it's 1, 3, 5, 7.

Anyway, I assumed 2 styles of circles used. With that assumption, we've got "mathematical circle" which is used for people, and "fuller corners" circle - used for your agents and for targets. If we use only these two styles, then the next size after the 5x5 you've pasted is the one from targets.

> New bug: It looks like the minimap doesn't update for the zoom I've set in `rules.ini` and just uses noweapon zoom instead.

Not sure what you mean, it it with the min equals max in config?

**Reply by mefistotelis** [reply]
> I've always welcomed a more zoomed out minimap, as it warns you earlier of enemy groups/vehicles closing in on your position and you don't have to move the camera around so much to scan the map for enemies at the start of a mission. Though I also don't use any abnormally high (low) zoom levels outside of the original game's highest (LR Rifle). This should optimally emulate the original game and have an .ini setting to adjust for personal preference.

I did some changes. I've set the default below original 640x480, but even more above original 320x200. Also exported the value to `rules.ini`.

While the new value is above what I would set it, I do agree higher values are highly useful to see enemies as they approach.
And most people played the game in 640x480, so that's the better mode to emulate.

**Reply by geist22** [reply]
> Not sure what you mean, it it with the min equals max in config?

I've set my zoom to be static (both min and max values the same), but the minimap shows default (noweapon) zoom instead of the one it should use. Minimap zoom does not match main screen zoom.
It seems minimap zoom is now only updated when you change weapons, but not at the start of a mission. 

Player unit circles are good now in 640x480 with Minigun+ zoom, but with Uzi and lower they change back to dotted circles, along with NPCs changing to squares. Maybe set a max size for this to prevent it?
